src/ledger/backend/connection.spec.ts

```diff
@@ -47,13 +48,28 @@ describe("ErdstallConnection", () => {
 	it("allows depositing into the erdstall contract", async () => {
 		const depositRegistered = EventHelper.within(10000, conn, "Deposited");
 
-		const stages = await conn.deposit(assets);
-		for (const stage of stages) {
-			const ctx = await stage.value;
-			const rec = await ctx.wait();
+		const [stages, numStages] = await conn.deposit(assets);
+		for await (const [name, stage] of stages) {
+			const rec = await stage.wait();
 			expect(rec.status, "depositing should have worked").to.equal(0x1);
 		}
 
+		let expectedNumStages: number = 0;
+		assets.values.forEach((value, key) => {
```

Comment:
I would prefer a more generic implementation which does not require previous knowledge about the type of contracts used, similar to this:

```ts
function depositStagesForAssets(a: Assets): number {
	return Array.from(a.values.entries())
		.map(([token, asset]) => {
			switch (asset.typeTag()) {
				case TypeTags.Amount:
					// ETHZERO -> one onchain transfer.
					// ERC20 -> one onchain approve + one onchain transfer.
					return token === ETHZERO ? 1 : 2;
				case TypeTags.Tokens:
					// ERC721 -> Tokens.length onchain approves + one onchain transfer.
					return (asset as Tokens).value.length + 1;
			}
		})
		.reduce((a, b) => a + b);
}
```

Adjust to your liking, but this way you do not have to bother with `addressKey`s at all, since `ETHZERO` is `0x00000...` which is the same no matter if checksum encoded or not.